src/main/java/lol/game/BattlefieldTraversal.java

```diff
@@ -45,4 +45,122 @@ public void visitAdjacent(int x, int y, int radius, TileVisitor visitor) {
       }
     }
   }
-}
\ No newline at end of file
+
+  public boolean noObstacles(int fromX, int fromY, int toX, int toY, Battlefield battlefield) {
+    int dx = Math.abs(toX - fromX);
+    int dy = Math.abs(toY - fromY);
+    if(dy <= 1 || dx <= 1){
+      return true;
+    }
+    else if(dy == dx){
+      if(!diagonalCheck(fromX, fromY, toX, toY, battlefield)) {
+        System.out.println("Obstacle on the way to the target of pos "+ toX +" "+ toY +" ! Can't shoot.");
+        return false;
+      }
+    }
+    else if(fromY == toY){
+      if(!verticalCheck(fromX, fromY, toX, toY, battlefield)) {
+        System.out.println("Obstacle on the way to the target of pos "+ toX +" "+ toY +" ! Can't shoot.");
+        return false;
+      }
+    }
+    else if(fromX == toX){
+      if(!horizontalCheck(fromX, fromY, toX, toY, battlefield)) {
+        System.out.println("Obstacle on the way to the target of pos "+ toX +" "+ toY +" ! Can't shoot.");
+        return false;
+      }
+    }
+    return true;
+  }
+
+
+  public boolean diagonalCheck(int x, int y, int toX, int toY, Battlefield battlefield) {
+    boolean emptyTile = battlefield.canPlaceAt(x, y);
+    System.out.println(x);
+    System.out.println(y);
+    System.out.println("------------------------");
+    if(x < toX && y < toY) {
+      for(int i = x; i < toX; i++) {
+        for(int j = y; j < toY; j++) {
+          if(!emptyTile) {
```

Comment:
`emptyTile` is never modified in the loop, so the loop are not useful? Same comment for the rest.

Reply:
I meant that according the position of the champion, the tiles to check would be different and some obstacles which are not on the way of the shoot could affect the truth value. That's why I created these 3 different methods. I don't know if it's clearer like that, it's hard to explain.

Reply:
What I meant is that since `emptyTile` is never modified in the loops, the following two codes are equivalent:

Code 1:
```java
for(int i = x; i < toX; i++) {
        for(int j = y; j < toY; j++) {
          if(!emptyTile) {
             return false;
```

and Code 2:

```java
if(!emptyTile) {
  return false;
}
```

Reply:
Yes that part is a mistake which I am currently fixing !

Reply:
Corrected the methods but there is a problem with the horizontal check which is really weird since it's basically the same method as the vertical one which is working well.
Otherwise everything does work !

Reply:
You could actually have the same code for diagonal / vertical / horizontal checks. Please look how it is done in `visitAdjacent`. By using an array of directions, it will make the code much less verbose.

Reply:
Also, you will need to pull from the main repository (see lab 5 instructions)

Reply:
Ahah ! I will try to do this today :)